src/main/java/software/amazon/smithy/lsp/SmithyTextDocumentService.java

```diff
@@ -157,9 +166,28 @@ private MessageParams msg(final MessageType sev, final String cont) {
         LspLog.println("Asking to complete " + position + " in class " + position.getTextDocument().getClass());
 
         try {
-            String found = findToken(position.getTextDocument().getUri(), position.getPosition());
+            String documentUri = position.getTextDocument().getUri();
+            String found = findToken(documentUri, position.getPosition());
+            DocumentPreamble preamble = documentPreamble(textBufferContents(documentUri));
+            LspLog.println("Import statements will be inserted at: " + preamble);
             LspLog.println("Token for completion: " + found + " in class " + position.getTextDocument().getClass());
-            return Utils.completableFuture(Either.forLeft(project.getCompletions(found)));
+
+            List<CompletionItem> items = project.getCompletions(found).stream().map(sci -> {
+                CompletionItem result = sci.getCompletionItem();
+                Optional<String> qualifiedImport = sci.getQualifiedImport();
+                Optional<String> importNamespace = sci.getImportNamespace();
+                if (!qualifiedImport.isEmpty() && !preamble.hasImport(qualifiedImport.get())
+                        && !importNamespace.equals(Optional.of(Constants.SMITHY_PRELUDE_NAMESPACE))) {
+                    TextEdit te = insertLine("use " + qualifiedImport.get(), preamble);
+                    result.setAdditionalTextEdits(ListUtils.of(te));
```

Comment:
This is the magic line - if the completion item qualifies (not imported, not smithy built-in) we modify the additionalTextEdits list with a newly constructed text edit.

TODO: Move to `Completions.java`